go-sdk/internal/sparse-merkle-tree/smt/merkletree.go

```diff
@@ -78,18 +80,33 @@ func (mt *sparseMerkleTree) AddLeaf(node core.Node) error {
 	// use up all the bits in the index's path. As soon as a unique path is found,
 	// which may be only the first few bits of the index, the new leaf node is added.
 	// One or more branch nodes may be created to accommodate the new leaf node.
-	newRootKey, err := mt.addLeaf(node, mt.rootKey, 0, path)
+	batch, err := mt.db.BeginBatch()
 	if err != nil {
 		return err
 	}
+	newRootKey, err := mt.addLeaf(batch, node, mt.rootKey, 0, path)
+	if err != nil {
+		log.L().Errorf("Error adding leaf node %s: %v, rolling back", node.Ref().Hex(), err)
+		_ = batch.Rollback()
+		return err
+	}
 	mt.rootKey = newRootKey
 
 	// update the root node index in the storage
-	err = mt.db.UpsertRootNodeIndex(mt.rootKey)
+	log.L().Infof("Upserting root node index to %s", mt.rootKey.Hex())
+	err = batch.UpsertRootNodeIndex(mt.rootKey)
 	if err != nil {
+		log.L().Errorf("Error upserting root node %s: %v, rolling back", mt.rootKey.Hex(), err)
+		_ = batch.Rollback()
+		return err
+	}
+	log.L().Infof("Committing batch operations for adding leaf node %s", node.Ref().Hex())
+	err = batch.Commit()
+	if err != nil {
+		log.L().Errorf("Error committing batch operations for adding leaf node %s: %v", node.Ref().Hex(), err)
+		_ = batch.Rollback()
 		return err
 	}
```

Comment:
Just an opinion - instead of relying on this format where the user need to remember to Commit() and RollBack() on error - I prefer this sort of style that Gorm offers 
```
db.Transaction(func(tx *gorm.DB) error {
  // do some database operations in the transaction (use 'tx' from this point, not 'db')
  if err := tx.Create(&Animal{Name: "Giraffe"}).Error; err != nil {
    // return any error will rollback
    return err
  }

  if err := tx.Create(&Animal{Name: "Lion"}).Error; err != nil {
    return err
  }

  // return nil will commit the whole transaction
  return nil
})
```

I've seen this pattern in multiple Golang based DB provider which flips it into: "tell me what work I need to do and I will managed it" instead of here "these are my functions, call me in the right order when you want to start doing work and finish doing work"

Reply:
I tried that style first, but found the convention (return `nil` to commit and `error` to rollback) awkward with nested calls to addLeaf() and extendPath(). on the other hand, the `batch.Commit()` and `batch.Rollback()` are more easier to follow especially when they (`Rollback()` in particular) need to be called in nested levels.

Reply:
okay makes sense